### ISSUE TITLE
Manifest enforces newest jGraphT version available

### ISSUE THREAD

**Issue by HannesWell** [jgrapht/jgrapht#736]
The Manifest of the JGraphT-core library itself imports the packages which it defines without any version restrictions (at least in the latest versions). When the respective bundle is required by an Eclipse plug-in and multiple versions of jGraphT-core are available, always the newest version is used even if the client plug-in specifies an older version.

Specifically, this problem occurs for me when jGraphT-core ver. 1.0.1 and ver. 1.3 are in the target platform. Some of my plug-ins require version 1.0.1 (specified in their Manifest) but nevertheless they use version 1.3. Since the API changed between 1.0.1 and 1.3 this is a serious problem.
The reason is the described self import (this was not easy to track down).

To clarify an edited excerpt of the Manifest of jGraphT-core version 1.0.1:
```
Import-Package: org.jgrapht,
org.jgrapht.alg
Export-Package: org.jgrapht;version="1.0.1",
org.jgrapht.alg;version="1.0.1"
```
As a workaround I modified the Manifest analogously to:
```
Import-Package: org.jgrapht;version="[1.0.1,1.1)",
org.jgrapht.alg;version="[1.0.1,1.1)"
Export-Package: org.jgrapht;version="1.0.1",
org.jgrapht.alg;version="1.0.1"
```
Using this modified Manifest the required version 1.0.1 finally is really used by client plug-ins. The jGraphT-core ver 1.0.1 bundle imports only its own packages because it is the only bundle within the range [1.0.1,1.1).

I didn't check the other jGraphT libraries but I think the same problem could occur there too.

I'm not really familiar with the purpose of importing own packages, but I assume the purpose is to ensure always the newest version is used by Eclipse. But as already stated, this is not a handy feature in case of API changes.
Therefore I want to suggest two options to solve the described problem:
1. Restrict the version of imported packages to API compatible versions. This correspondences to my workaround.
2. Drop the import of own packages (only if it is to ensure the newest version). Since client plug-ins can broadly specify the version they want to use and _always the newest_ is also possible, import of own packages is not necessary in my opinion (only in case this is the purpose of self import).

**Comment by HannesWell** [comment]
I found a solution how to specify the packages imported in the MANIFEST.MF at the Apache Felix Maven Bundle Plugin documentation:
https://felix.apache.org/documentation/subprojects/apache-felix-maven-bundle-plugin-bnd.html#import-package

The subsequent [_default behaviour_ section](https://felix.apache.org/documentation/subprojects/apache-felix-maven-bundle-plugin-bnd.html#default-behavior)  states that:
> < Import-Package > is assumed to be "*", which imports everything referred to by the bundle content, but not contained in the bundle. _Any exported packages are also imported by default, to ensure a consistent class space._
Which explains why all jgrapht-packages are imported again.

From the discussion [if a bundle should import its exported package](https://felix.apache.org/documentation/tutorials-examples-and-presentations/apache-felix-osgi-faq.html#should-a-bundle-import-its-own-exported-packages), I conclude that JGraphT is one of the exceptions where export only is wanted, since it is a pure library.
Furthermore use of different versions of JGraphT from one Eclipse target-platform is not possible otherwise, because of the missing version constraints. Additionally the _one-version-backwards compatibility_ policy of JGraphT combined with not specified version number increments (sometimes minor, sometimes patch) makes it impossible to specify a valid version range for the next version in advance. Therefore I think it is better to drop the self-import.

I created a ready to submit [branch in my JGraphT clone](https://github.com/HannesWell/jgrapht/tree/avoidSelfImport), where I added the following lines to the `org.apache.felix`-plugin element in the pom.xml of each module
```
    <configuration>
        <instructions>
            <Import-Package>!org.jgrapht.*,*</Import-Package>
        </instructions>
    </configuration>
```
The import first excludes all org.jgrapht packages from the Imported-packages and then adds all remaining required/referenced packages. Therefore if new `org.jgrapht`-packages or new imports/referred packages are added this section does not need to be adjusted.

If you agree with this change, I can create a PR.

**Comment by d-michail** [comment]
I am not sure I understand the implications of this change. Is there no workaround for your use-case?

I would prefer to avoid, doing a change which I do not fully understand.  @jsichi Any feedback on this? Have you encountered such issues before?

**Comment by jsichi** [comment]
I only have a tiny bit of experience with OSGi, and most of it involved me cursing.  Given the description of the problem with multiple JGraphT versions required by different components, and the documentation on libraries cited by Hannes, it sounds like we should make the change to eliminate self-importing.

@HannesWell can you prepare the PR so we can take a look?

**Comment by HannesWell** [comment]
In our scenario, we had two versions of JGraphT in the Target-Platform of our Eclipse-RCP-Application that where located in one or two (I don't remember exactly) p2-repositories and multiple Eclipse Plug-ins that reference to jGraphT in their MANIFEST.MF via `Require-Bundle`. The only workaround I found was to add the version-constraint to the MANIFEST.MF in the jars by hand, as described in the initial comment. I think this isn't the way to go.

The implications for the jar-build are, that the `Import-Package` section in the generated Manifest contains only those required packages that are not contained in and exported by this jar. I checked this for jgrapht-core by building the jar with and without my proposed change, unzipped them and run a diff on the resulting directories. The only relevant difference was the `Import-Package` section of the MANIFEST.MF (additionally the build-timestamp and the packed pom.xml had obvious differences).

Regarding the implications for the use of the jars in a OSGI-environment: I also only have basic 'user'-knowledge of OSGI and I'm not the expert in it. But as far as I understand it from the linked discussion [if a bundle should import its exported package](https://felix.apache.org/documentation/tutorials-examples-and-presentations/apache-felix-osgi-faq.html#should-a-bundle-import-its-own-exported-packages), the only purpose of importing the own packages is that the number of active versions of a bundle (in this case jgrapht) can be reduced. But since JGraphT is a pure library and API breaks may be done between different versions, this behaviour is not wanted. Especially since Eclipse/OSGI here imports the newest version of a bundle/package available , which forces plug-ins/bundles referencing JGraphT via `Require-Bundle` to the newest version.

I don't know exactly why importing and exporting a package makes that imported (newer) package accessible to referencing Eclipse Plug-ins, but obviously OSGI loops those packages through, even the export has a version constraint? It's strange for me.
Maybe because by importing and then exporting a package through a bundle this implicitly redeclares the package-version to be the version of the bundle?

I also had a look at the [pom.xml of Google's guava library](https://github.com/google/guava/blob/ac0c3424411c0974010d754042d75e9d82665379/guava/pom.xml#L81) and they even explicitly name their imports and do not import their own packages. I see this as an indication that importing own packages isn't a necessary feature.

@jsichi Yes, I will create a PR right now.
@d-michail It looks like your jheaps library has the same issue. If you want I can open a PR there too?
